scripts/runnable_recipe_ray_vllm_weight_sync.py

```diff
@@ -368,6 +369,447 @@ def run(self):
             idx += 1
 
 
+# =============================================================================================
+
+from typing import Dict, Any
+from vllm import LLM, SamplingParams
+import vllm
+import torch
+from tensordict import TensorDictBase, TensorClass, TensorDict, NonTensorStack, lazy_stack, from_dataclass
+from tensordict.utils import _zip_strict, expand_as_right
+
+# class Trajectory(TensorClass):
+#     text: str
+#     tokens: torch.Tensor|None=None
+#     response: str | None=None
+#     # response_tokens --> tokens_response because Transform hardcodes..
+#     tokens_response: torch.Tensor | None=None
+#     # prompt_attention_mask: torch.Tensor | None=None
+#     log_probs: torch.Tensor | None=None
+
+def vllm_generate(
+    tensordict: TensorDictBase | Trajectory,
+    vllm_instance,  # The LLM object
+    generate_kwargs: Dict[str, Any] | None = None,
+    tokenizer = None,
+    text_key: str = "text",
+    token_key: str = "tokens",
+    # response_tokens --> tokens_response because Transform hardcodes..
+    token_response_key: str = "tokens_response",
+    text_response_key: str = "text_response",
+    log_prob_key: str = "log_probs",
+    attention_mask_key: str = "attention_mask",
+    padding_value: int = -1,
+) -> Trajectory:
+    
+    args = ()
+    
+    if generate_kwargs is None:
+        generate_kwargs = {}
+    generate_kwargs.setdefault("detokenize", True)
+    generate_kwargs.setdefault("prompt_logprobs", False)
+    generate_kwargs.setdefault("logprobs", True)
+    # Create SamplingParams from generate_kwargs
+    sampling_params = SamplingParams(**generate_kwargs)
+    kwargs = {"sampling_params": sampling_params, "use_tqdm": False}
+
+    
+    txt = tensordict.get(text_key)
+    if not isinstance(txt, (list, str)):
+        txt = txt.tolist()
+    args = (txt,)
+
+    tokens_out = vllm_instance.generate(*args, **kwargs)
+    tokens_out = _get_output_tokens_and_log_probs(tokens_out, tokenizer, log_prob_key, token_response_key, text_response_key, padding_value)
+
+    in_keys = [
+        log_prob_key,
+        token_response_key,
+        text_response_key,
+        token_key,
+        attention_mask_key,
+    ]
+    out = tokens_out.select(*in_keys, strict=False)
+
+    # ================================================================
+    out[token_key] = tensordict[token_key]
+    # ================================================================
+
+    inputs = tensordict.select(text_key, strict=False)
+    if inputs.ndim < out.ndim:
+        inputs = inputs.unsqueeze(-1).expand(out.shape)
+    out.update(inputs)
+
+    td = tensordict.clone()
+    td.update(out, keys_to_update=list(out.keys()))
+
+    # This runs into errors with accessing resut["bla"] e.g. llm.py:511
+    # return Trajectory.from_tensordict(out)
+    return td 
+
+
+CompletionOutput_tc = from_dataclass(vllm.outputs.CompletionOutput)
+
+class _RequestOutput_tc(TensorClass["nocast"]):
+    request_id: str
```

Comment:
@vmoens Connected to our chat this morning, I would type these as `Batched[string]` or `Batched[List[str]]`. In the first case, I would interpret `Batched` as "Some iterable of len bsz".

Also, this should have a batch size, no?